datafusion/core/src/execution/context/mod.rs

```diff
@@ -2179,6 +2182,21 @@ impl FunctionRegistry for SessionState {
     fn deregister_udwf(&mut self, name: &str) -> Result<Option<Arc<WindowUDF>>> {
         Ok(self.window_functions.remove(name))
     }
+
+    fn deregister_analyzer_rule(&mut self, index: usize) -> Result<()> {
```

Comment:
If we added `SessionContext::analyzer_rules` I don't think this API is necessary -- instead the user could just override the existing list.